xapian-letor/api/featurelist_internal.cc

```diff
@@ -0,0 +1,177 @@
+	Xapian::doccount df = featurelist_db.get_termfreq(*qt);
+	idf[*qt] = (df == 0) ? 0 : log10((double)totaldocs / (double)(1 + df));
+    }
+    std::swap(inverse_doc_freq, idf);
+}
+
+void
+FeatureList::Internal::compute_doc_length()
+{
+    std::map<std::string, long int> len;
+
+    long int temp_count = 0;
+    Xapian::TermIterator dt = featurelist_doc.termlist_begin();
+    // reach the iterator to the start of the title terms i.e. prefix "S"
+    dt.skip_to("S");
+    for ( ; dt != featurelist_doc.termlist_end(); ++dt) {
+	if ((*dt)[0] != 'S') {
+	    // We've reached the end of the S-prefixed terms.
+	    break;
+	}
+	temp_count += dt.get_wdf();
+    }
+    len["title"] = temp_count;
+    len["whole"] = featurelist_db.get_doclength(featurelist_doc.get_docid());
+    len["body"] = len["whole"] - len["title"];
+    std::swap(doc_length, len);
+}
+
+void
+FeatureList::Internal::compute_collection_length()
+{
+    std::map<std::string, long int> len;
+
+    if (!featurelist_db.get_metadata("collection_len_title").empty() &&
+	    !featurelist_db.get_metadata("collection_len_body").empty() &&
+	    !featurelist_db.get_metadata("collection_len_whole").empty()) {
+	len["title"] = atol(featurelist_db.get_metadata("collection_len_title").c_str());
+	len["body"] = atol(featurelist_db.get_metadata("collection_len_body").c_str());
+	len["whole"] = atol(featurelist_db.get_metadata("collection_len_whole").c_str());
+    } else {
+	long int temp_count = 0;
+	Xapian::TermIterator dt = featurelist_db.allterms_begin("S");
+	for ( ; dt != featurelist_db.allterms_end("S"); ++dt) {
+	    //  because we don't want the unique terms so we want their
+	    // original frequencies and i.e. the total size of the title collection.
+	    temp_count += featurelist_db.get_collection_freq(*dt);
+	}
+	len["title"] = temp_count;
+	len["whole"] = featurelist_db.get_avlength() * featurelist_db.get_doccount();
+	len["body"] = len["whole"] - len["title"];
+    }
+    std::swap(collection_length, len);
+}
+
+void
+FeatureList::Internal::compute_collection_termfreq()
+{
+    std::map<std::string, long int> tf;
+
+    for (Xapian::TermIterator qt = featurelist_query.get_unique_terms_begin();
+	 qt != featurelist_query.get_terms_end(); ++qt) {
+	tf[*qt] = featurelist_db.get_collection_freq(*qt);
+    }
+    std::swap(collection_termfreq, tf);
+}
+
+void
+FeatureList::Internal::populate_feature(Feature *feature_)
+{
+    stat_flags stats_needed = stat_flags(feature_->get_stats());
+    if (stats_needed & TERM_FREQUENCY)
+	feature_->set_termfreq(termfreq);
+    if (stats_needed & INVERSE_DOCUMENT_FREQUENCY)
+	feature_->set_inverse_doc_freq(inverse_doc_freq);
+    if (stats_needed & DOCUMENT_LENGTH)
+	feature_->set_doc_length(doc_length);
+    if (stats_needed & COLLECTION_LENGTH)
+	feature_->set_collection_length(collection_length);
+    if (stats_needed & COLLECTION_TERM_FREQ)
+	feature_->set_collection_termfreq(collection_termfreq);
```

Comment:
Clearing such a map will indeed involve O(N) work because the strings need to be released.  However, destroy the map will also be O(N) - this is not an additional cost, just a change in the timing of when it is incurred - at the moment this happens when the local map is destroyed after that `swap()`.

In fact by being lazier here we may be able to avoid that cost, because we avoid filling any of the maps which aren't going to be used.  And if we don't fill them then N is 0.

FWIW, I'm OK with leaving this as a follow-on change after we have this merged - the code as-is should work OK, but could be optimised.